### ISSUE TITLE
Unable to get the exact same outputs by passing Z vectors to the decoder network in an autoencoder with batch normalization layers

### ISSUE THREAD

**Issue by Amir-Arsalan** [torch/torch7#687]
It's been more than a day that I've been trying to find out why my auto-encoder is not giving me the same output when I use a pre-stored Z vector. I did my best by checking all my code and I think my code is fine and this is a Torch problem. So I thought maybe I should submit an issue. I tried to simplify the problem as much as possible and bring it down to a simple forward pass on a [fully-convolutional] auto-encoder without any training and try it with random data. I May note that I did not have this issue with fully-connected models.

Here's my first observation:
1- Do a forward pass for example(s) 'X'
2- Store the 'Z' vector(s)
3- Store the output(s) 'O'
4- Pass the 'Z' vector(s) to the decoder
5- Obtain the output(s) 'O^'
6- Take the sum of absolute pixel-wise differences between 'O' and 'O^'
7- The sum is not 0!

Here's another observation:
1- Do a forward pass for example(s) 'X'
2- Store the output(s) 'O'
3- Do a forward pass on the decoder using only some of the previous 'Z' vector(s) obtained through the last forward pass
4- Obtain the output(s) 'O^' (the output(s) MUST be exactly the same with 'O')
6- Take the sum of absolute pixel-wise differences between 'O' and 'O^'
7- The sum is not 0! (Though less than the previous observation)

And finally:
1- Do a forward pass for example(s) 'X'
2- Store the output(s) 'O'
3- Do a forward pass on the decoder using ALL of the previous 'Z' vector(s) obtained through the last forward pass
4- Obtain the output(s) 'O^' (the output(s) MUST be exactly the same with 'O')
6- Take the sum of absolute pixel-wise differences between 'O' and 'O^'
7- The sum IS 0!

I thought the issue is a precision problem but I ran my code on GPU and CPU (single and double precision) and still the output results are different. Can someone give me a reason on why this is happening and whether I am doing something wrong?

Side note: After some training epochs, the resultant image(s) obtained after passing pre-stored Z vector(s) and the resultant image(s) obtained after passing some of the network's own Z vectors become visually visible. The deviation becomes more easy to visually notice when passing all of the network's own Z vectors. I uploaded two of such examples in a repository I made which you can access to through the link below along with the code used:
https://github.com/Amir-Arsalan/Torch_Test

**Comment by Amir-Arsalan** [comment]
@soumith
I just commented the lines that add the batch normalization layer to the decoder network and the problem gets resolved. There is something happening in the the batch normalization layers causing the issue. Can you double check on this?

**Comment by soumith** [comment]
In the future, could you post this to our google groups here: https://groups.google.com/forum/#!forum/torch7 , we only use this github issues for bug tracking the torch7 package. We'll happily reply there. Thanks.

**Comment by Amir-Arsalan** [comment]
@soumith  I think one easy way to resolve this issue is to only update the mean and variances when the gradInput changes so that the network will be in the test mode by default unless a backward pass is called. What do you think?
